### DISCUSSION TITLE
specification of the `foreach` statement (12.9.5) doesn't apply for `System.Span`

### DISCUSSION THREAD

**Discussion by ltcmelo** [dotnet/csharplang#6288] [General]
The [`foreach` statement](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/statements#1295-the-foreach-statement) is specified by means of an expansion into a `while` *.  
But the specification doesn't apply for a type such as `System.Span` or `System.ReadOnlySpan` as the expansion will produce invalid code.  
Specifically: the `V` in `foreach (V v in x) «embedded_statement»` will expand into the `V` in `V v = (V)(T)e.Current;`; however, if the `foreach` is inside an `async` method, then the declaration `V v` will lead to a diagnose such as:

`error CS4012: Parameters or locals of type 'ReadOnlySpan<char>' cannot be declared in async methods or async lambda expressions.` 

Here's an example in which specified expansion breaks.

```
public class C
{
    public async Task M()
    {
        IReadOnlyList<string> o = null;
        foreach (ReadOnlySpan<char> c in o) {}        
        return;
    }
}
```

\* I have (most of) this expansion as a [syntax rewriter](https://github.com/ShiftLeftSecurity/SharpSyntaxRewriter/blob/master/src/SharpSyntaxRewriter/Rewriters/ExpandForeach.cs).

**Comment by CyrusNajmabadi** [comment]
@ltcmelo It's unclear what you're asking for here.

**Reply by ltcmelo** [reply]
I'm reporting a bug in the specification.

**Reply by ltcmelo** [reply]
I suppose that I need (to click on the button) "Create issue from discussion" then.

**Reply by CyrusNajmabadi** [reply]
> I'm reporting a bug in the specification.

I don't understand what the bug is that you're reporting. Can you expand on this?

**Reply by BillWagner** [reply]
@ltcmelo 

C# specification (standard) issues should be reported in the [`dotnet/csharpstandard`](https://github.com/dotnet/csharpstandard) repo.  That's where the committee is doing its work. This will be a timely issue. The committee is working on the updates for C# 7.x and we would have likely missed that `ref struct` declarations would impact this area of the language. (Its less about `foreach`, but more relate to any use of a `ref struct` in an `async` method.

**Reply by ltcmelo** [reply]
OK, will do that then. Thanks.

**Reply by ltcmelo** [reply]
I was told that the root of the issue is Roslyn accepting the original program, and that I should open a report in Roslyn. Here it is: https://github.com/dotnet/roslyn/issues/62747
